rand_distr/src/zipf.rs

```diff
@@ -0,0 +1,322 @@
+// Copyright 2021 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! The Zeta and related distributions.
+
+use num_traits::Float;
+use crate::{Distribution, Standard};
+use rand::{Rng, distributions::OpenClosed01};
+use core::fmt;
+
+/// Samples floating-point numbers according to the zeta distribution.
+///
+/// The zeta distribution is a limit of the [`Zipf`] distribution.
+///
+/// # Example
+/// ```
+/// use rand::prelude::*;
+/// use rand_distr::Zeta;
+///
+/// let val: f64 = thread_rng().sample(Zeta::new(1.5).unwrap());
+/// println!("{}", val);
+/// ```
+#[derive(Clone, Copy, Debug)]
+pub struct Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    a_minus_1: F,
+    b: F,
+}
+
+/// Error type returned from `Zeta::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ZetaError {
+    /// `a <= 1` or `nan`.
+    ATooSmall,
+}
+
+impl fmt::Display for ZetaError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            ZetaError::ATooSmall => "a <= 1 or is NaN in Zeta distribution",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+impl std::error::Error for ZetaError {}
+
+impl<F> Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    /// Construct a new `Zeta` distribution with given `a` parameter.
+    #[inline]
+    pub fn new(a: F) -> Result<Zeta<F>, ZetaError> {
+        if !(a > F::one()) {
+            return Err(ZetaError::ATooSmall);
+        }
+        let a_minus_1 = a - F::one();
+        let two = F::one() + F::one();
+        Ok(Zeta {
+            a_minus_1,
+            b: two.powf(a_minus_1),
+        })
+    }
+}
+
+impl<F> Distribution<F> for Zeta<F>
+where F: Float, Standard: Distribution<F>, OpenClosed01: Distribution<F>
+{
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> F {
+        // This is based on the numpy implementation.
+        loop {
+            let u = rng.sample(OpenClosed01);
+            let v = rng.sample(Standard);
+            let x = u.powf(-F::one() / self.a_minus_1).floor();
+
+            if x < F::one() {
```

Comment:
This step is not part of the original algorithm ([[Devroye86](https://doi.org/10.1007/978-1-4613-8643-8)] Section 6.1, page 551). Looking at the comments in the [numpy implementation](https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/random/src/distributions/distributions.c#L974), they say:
> The real result may be above what can be represented in a signed long. Since this is a straightforward rejection algorithm, we can just reject this value. This function then models a Zipf distribution truncated to sys.maxint.

This is related to the effective domain (which is bounded) for the implementation. (see documentation comment)

I personally do not see yet where is the problem with `x < 1`.

Reply:
On second thought, is there not something really disturbing about this? I mean, this condition should never be satisfied, right? 
- `u` is `OpenClosed01`, ie in (0, 1]
- Therefore, `1/u` is in [1, infty)
- Then, `(1 / u)^(1 / (a - 1))` is also in [1, infty)
- Finally, `x = floor( (1 / u)^(1 / (a - 1)) )` should always be in 1, 2, ...

Did I get something wrong?

Reply:
`x` could only be smaller than 1 if there is a rounding issue. I'll try to turn it into an assert instead.

Reply:
In that sense, do we want to do something for `x = f64::INFINITY`? 

This can happen because before applying `floor` the number is infinity, but it should fall back to `f64::MAX`. This is also a rounding issue, no? (I know it is an edge-case)